arbos/multigasCollector/collector.go

```diff
@@ -0,0 +1,308 @@
+// gas data from blocks. It receives BlockTransactionMultiGas data through a channel, buffers
+// it in memory, and periodically writes batches to disk in protobuf format.
+type Collector struct {
+	stopwaiter.StopWaiter
+
+	config Config
+	input  <-chan *CollectorMessage
+
+	batchNum          int64
+	blockBuffer       []*proto.BlockMultiGasData
+	transactionBuffer []*proto.TransactionMultiGasData
+}
+
+// ToProto converts the TransactionMultiGas to its protobuf representation.
+func (tx *TransactionMultiGas) ToProto() *proto.TransactionMultiGasData {
+	multiGasData := &proto.MultiGasData{
+		Computation:   tx.MultiGas.Get(multigas.ResourceKindComputation),
+		StorageAccess: tx.MultiGas.Get(multigas.ResourceKindStorageAccess),
+		StorageGrowth: tx.MultiGas.Get(multigas.ResourceKindStorageGrowth),
+		HistoryGrowth: tx.MultiGas.Get(multigas.ResourceKindHistoryGrowth),
+	}
+
+	if unknown := tx.MultiGas.Get(multigas.ResourceKindUnknown); unknown > 0 {
+		multiGasData.Unknown = &unknown
+	}
+
+	if refund := tx.MultiGas.GetRefund(); refund > 0 {
+		multiGasData.Refund = &refund
+	}
+
+	return &proto.TransactionMultiGasData{
+		TxHash:   tx.TxHash,
+		TxIndex:  tx.TxIndex,
+		MultiGas: multiGasData,
+	}
+}
+
+// ToProto converts the BlockInfo to its protobuf representation.
+func (btmg *BlockInfo) ToProto() *proto.BlockMultiGasData {
+	return &proto.BlockMultiGasData{
+		BlockNumber:    btmg.BlockNumber,
+		BlockHash:      btmg.BlockHash,
+		BlockTimestamp: btmg.BlockTimestamp,
+	}
+}
+
+// NewCollector returns an initialized collector.
+//
+// It validates the configuration and ensures the output directory exists.
+//
+// Parameters:
+//   - config: Configuration specifying output directory and batch size
+//   - input: Channel supplying CollectorMessage values
+//
+// Returns:
+//   - *Collector: The initialized collector
+//   - error: Configuration validation or initialization error
+//
+// Usage:
+//
+//	c, _ := NewCollector(cfg, input)
+//	c.Start(ctx)
+//	// ... send messages ...
+//	c.StopAndWait() // flushes and stops regardless of channel state
+func NewCollector(config Config, input <-chan *CollectorMessage) (*Collector, error) {
+	if config.OutputDir == "" {
+		return nil, ErrOutputDirRequired
+	}
+
+	if config.BatchSize == 0 {
+		return nil, ErrBatchSizeRequired
+	}
+
+	if err := os.MkdirAll(config.OutputDir, 0755); err != nil {
+		return nil, ErrCreateOutputDir
+	}
+
+	return &Collector{
+		config:            config,
+		input:             input,
+		batchNum:          0,
+		blockBuffer:       make([]*proto.BlockMultiGasData, 0, config.BatchSize),
+		transactionBuffer: make([]*proto.TransactionMultiGasData, 0, defaultTxPrealloc),
+	}, nil
+}
+
+// Start begins background processing using StopWaiter.
+// Processing continues until StopAndWait() is called or ctx is canceled.
+// Channel close is handled (remaining data is flushed), but shutdown does not
+// depend on it.
+func (c *Collector) Start(ctx context.Context) {
```

Comment:
How did you decide to make the Collector a StopWaiter (as opposed to just letting it's `ProcessData` method be called by a thread in the ExecutionEngine StopWaiter?)

I ask because I vaguely remember Lee not being a fan of nested StopWaiters, or at least, he didn't like the way they were nested in the BoLD codebase: https://linear.app/offchain-labs/issue/NIT-2974/fix-bold-stopwaiter-usage

But, maybe that was a different pattern. I'm really not saying it's wrong to do it your way. I'm just trying to understand how you were reasoning about it. It might be perfectly reasonable.

Reply:
We discussed it on a call, however since the collector is started from `ExecutionEngine` which is already StopWaiter and because we are moving channel into internals i'll remove it